src/thunderbots/software/backend_output/grsim/motion_controller.cpp

```
@@ -0,0 +1,182 @@
+    // and the change in time since the last motion controller run
+    double delta_speed_x, delta_speed_y, delta_angular_speed;
+
+    // check for negative sqrt case
+    if (pow(robot.velocity().len(), 2) <= 2 * ROBOT_MAX_ACCELERATION * distance_to_dest)
+    {
+        // if the sqrt is negative, the final speed will be negative (opposite direction
+        // of current speed)
+        expected_final_speed = -1 * sqrt(2 * ROBOT_MAX_ACCELERATION * distance_to_dest -
+                                         pow(robot.velocity().len(), 2));
+        b_can_stop_in_time   = true;
+    }
+    else
+    {
+        // calculate the expected final speed assuming max decceleration
+        expected_final_speed = sqrt(pow(robot.velocity().len(), 2) -
+                                    2 * ROBOT_MAX_ACCELERATION * distance_to_dest);
+
+        // the robot can stop in time if it's desired final speed is higher than the speed
+        // if the robot maximum decelerates from the current state
+        b_can_stop_in_time = expected_final_speed <= desired_final_speed;
+    }
+
+    // check negative sqrt case for angular speed
+    if (pow(robot.angularVelocity().toRadians(), 2) <=
+        fabs(2 * ROBOT_MAX_ANG_ACCELERATION * angle_to_dest))
+    {
+        expected_final_ang_speed =
+            -1 * sqrt(fabs(2 * ROBOT_MAX_ANG_ACCELERATION * angle_to_dest) -
+                      pow(robot.angularVelocity().toRadians(), 2));
+        b_can_stop_rotate_in_time = true;
+    }
+    else
+    {
+        expected_final_ang_speed  = sqrt(pow(robot.angularVelocity().toRadians(), 2) -
+                                        2 * ROBOT_MAX_ANG_ACCELERATION * angle_to_dest);
+        b_can_stop_rotate_in_time = expected_final_ang_speed <= 0.0;
+    }
+
+
+    if (delta_time < 0)
+    {  // there should be no negative changes in time except from the first run when
+       // variables are initialized.
+        delta_time = fabs(delta_time);
```

I might be wrong, but I think there should never be negative time, and if there is we should probably throw an exception here. At the very least we should add a `TODO` to throw an exception here (with a reference to #16) and just `assert(false)` so that we fail instead of just carrying on with invalid inputs (until #16 is finished anyhow). This might be a bit controversial though, @MathewMacDougall ?

---

Agree we shouldn't accept negative time. Probably best if this check goes at the top of the function. I think in this case because this is for the simulator, we can be slightly more tolerant of a negative value and simply return early (perhaps with the Robot's current trajectory, so no speed change) and log an error. We can always escalate this to an assert or exception if necessary.

---

Sounds good, do we currently have a logger in the new code? (Aside from the issue that talks abut a logger)

---

No logger in the current code (hence the ticket for it). For now just use `ROS_WARN` or similar, and be sure to add a comment linking to #33 so we know to replace it in the future.

---

Issue #123 created for error handling of negative delta_time.